subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -716,11 +730,43 @@ static void scheduler_srv_reset(struct bt_mesh_model *model)
 	srv->status_bitmap = 0;
 	k_delayed_work_cancel(&srv->delayed_work);
 	net_buf_simple_reset(srv->pub.msg);
+
+	memset(&srv->sch_reg, 0, sizeof(srv->sch_reg));
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		for (int idx = 0; idx < BT_MESH_SCHEDULER_ACTION_ENTRY_COUNT;
+		     ++idx) {
+			store(srv, idx);
```

Comment:
I think that to remove the data, `data` and `data_len` should be zero: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/mesh/access.c#L1158-L1162

Reply:
+1

Also, the memset above isn't strictly necessary, since the bitmap is the decision point

Reply:
memset is needed. If a scheduler is run after reset, it can wrongly set bitmap based on old register values in the memory.

Reply:
I didn't quite get that. If the bitmap is false for a scheduler entry, that entry shouldn't be accessed all, should it? The timer handler should be checking this, no?

Reply:
I guess it is worth to store bitmap as well and rely on bitmap when restore entries.

Reply:
> I guess it is worth to store bitmap as well and rely on bitmap when restore entries.

I think this is duplicate state

Reply:
> > I guess it is worth to store bitmap as well and rely on bitmap when restore entries.
> 
> I think this is duplicate state

otherwise memset is mandatory

Reply:
In that case, I don't understand what the bitmap does. Doesn't the bitmap represent which entries are active? If it does, we need to set the bit when entries are loaded, and the bitmap will be populated properly by the end of the load procedure. We're checking this bit when finding the first scheduleable entry

Reply:
Hold on. Before you diverge into some other discussion. The bitmap logic is not fully correct and this needs to be fixed separately. It is not related to the register storage and restore mechanism. See NCSDK-9281.

Reply:
> In that case, I don't understand what the bitmap does. Doesn't the bitmap represent which entries are active? If it does, we need to set the bit when entries are loaded, and the bitmap will be populated properly by the end of the load procedure. We're checking this bit when finding the first scheduleable entry

Yep. Bitmap represents active entries.
Loaded time doesn't mean the correct time. To make entry active it should pass through time revising procedures.
If time is expired the revise system will reject this entry and bit will never be set for it. 
If entry is periodical or not expired yet then the revise system should recalculate TAI time (gotten from Time server since time Zone or summer\winter time have been changed or etc) after this bit will be set.
And only after this the entry could be scheduled by looking at this bitmap as this is actual.

Reply:
I see, so I didn't understand what the bitmap was doing :smile: The bitmap shouldn't be stored then, right? None of the entries would be active until we're able to recover the current time using a time status message from the mesh